### PR TITLE
[VEN-105] Product states - pre-Orders support - drop available_on logic

### PR DESCRIPTION
**Pull request by kgorazd**

### CODE REVIEW

File: api/app/controllers/spree/api/v2/platform/orders_controller.rb [hunk 1]
```diff
@@ -100,7 +100,7 @@ def model_class
           end
 
           def allowed_sort_attributes
-            super.push(:available_on, :total, :payment_total, :item_total, :shipment_total,
+            super.push(:make_active_at, :total, :payment_total, :item_total, :shipment_total,
```
Comment (damianlegawiec) [inline api/app/controllers/spree/api/v2/platform/orders_controller.rb:103]: @kgorazd let's not change the name of `available_on`

File: core/db/migrate/20220103082046_add_status_and_make_active_at_to_spree_products.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class AddStatusAndMakeActiveAtToSpreeProducts < ActiveRecord::Migration[5.2]
+  def change
+    add_column :spree_products, :status, :string, null: false, default: 'draft'
+    add_index :spree_products, :status
+    rename_column :spree_products, :available_on, :make_active_at
```
Comment (damianlegawiec) [inline core/db/migrate/20220103082046_add_status_and_make_active_at_to_spree_products.rb:5]:
```suggestion
```

File: core/db/migrate/20220103082046_add_status_and_make_active_at_to_spree_products.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class AddStatusAndMakeActiveAtToSpreeProducts < ActiveRecord::Migration[5.2]
+  def change
+    add_column :spree_products, :status, :string, null: false, default: 'draft'
+    add_index :spree_products, :status
```
Comment (damianlegawiec) [inline core/db/migrate/20220103082046_add_status_and_make_active_at_to_spree_products.rb:4]:
```suggestion
    add_index :spree_products, :status
    add_index :spree_products, %i[status deleted_at]
```

File: core/lib/tasks/core.rake [hunk 4]
```diff
@@ -185,3 +185,19 @@ use rake db:load_file[/absolute/path/to/sample/filename.rb]}
     end
   end
 end
+
+namespace :core do
+  desc 'Set "active" status on draft products where available_on is in the past'
+  task activate_products: :environment do |_t, _args|
+    Spree::Product.where('available_on <= ?', Time.current).where(status: 'draft').find_each do |product|
+      product.update_attribute(:status, 'active')
+    end
```
Comment (damianlegawiec) [inline core/lib/tasks/core.rake:194]:
```suggestion
    Spree::Product.where('available_on <= ?', Time.current).where(status: 'draft').update_all(status: 'active', updated_at: Time.current)
```

File: core/lib/tasks/core.rake [hunk 5]
```diff
@@ -185,3 +185,19 @@ use rake db:load_file[/absolute/path/to/sample/filename.rb]}
...
+  end
+
+  desc 'Set "archived" status on active products where discontinue_on is in the past'
+  task archive_products: :environment do |_t, _args|
+    Spree::Product.where('discontinue_on <= ?', Time.current).where.not(status: 'archived').find_each do |product|
```
Comment (damianlegawiec) [inline core/lib/tasks/core.rake:199]: @kgorazd same here

File: api/app/serializers/spree/v2/storefront/product_serializer.rb [hunk 6]
```diff
@@ -6,7 +6,7 @@ class ProductSerializer < BaseSerializer
 
         set_type :product
 
-        attributes :name, :description, :available_on, :slug, :meta_description, :meta_keywords, :updated_at, :sku, :public_metadata
+        attributes :name, :description, :available_on, :status, :slug, :meta_description, :meta_keywords, :updated_at, :sku, :public_metadata
```
[the remark this answers is not in the document]
Comment (kgorazd) [reply api/app/serializers/spree/v2/storefront/product_serializer.rb]: done

File: core/db/migrate/20220103082046_add_status_and_make_active_at_to_spree_products.rb (hunk 3 above)
Comment (kgorazd) [reply core/db/migrate/20220103082046_add_status_and_make_active_at_to_spree_products.rb:4]: done

File: core/lib/tasks/core.rake (hunk 4 above)
Comment (kgorazd) [reply core/lib/tasks/core.rake]: done

File: core/lib/tasks/core.rake (hunk 5 above)
Comment (kgorazd) [reply core/lib/tasks/core.rake]: done

**Review by damianlegawiec — APPROVED** [review]
Awesome work @kgorazd 👍

**Pull request merged**